File: 2025/Day3/python/D3PartTwo.py
def read_file(filename:str = "../sample.txt") -> str:
    with open(filename, 'r') as f:
        d = f.read()
    return d

def sum_total(a:list[int]) -> int:
    total = 0
    for num in a:
        total *= 10
        total += num
    return total

def solve(line:str, battery_banks:int = 12) -> int:
    res = []
    for ch in line:
        if len(res) == 0:
            res.append(int(ch))
            continue
        if int(ch) >= min(res) and len(res) + 1 > battery_banks:
            temp = res.copy()
            temp.remove(min(temp))
            temp.append(int(ch))
            if sum_total(temp) > sum_total(res[1:] + [int(ch)]):
                res.remove(min(res))
            else:
                res.remove(res[0])
            res.append(int(ch))
        elif len(res) < battery_banks:
            res.append(int(ch))
    total = 0
    for num in res:
        total *= 10
        total += num
    print(line, total, res)
    return total

filename = "../input.txt"
# file_data = read_file(filename)
file_data = read_file()
total = 0
for i in range(len(file_data.split('\n'))):
    line = file_data.split('\n')[i]
    total += solve(line)
print("TOTAL:", total)
print("\nDIFF:", 3121910778619 - total)
assert total == 3121910778619

file_data = read_file(filename=filename).split('\n')
if file_data[-1] == "":
    del file_data[-1]
total = 0
for i in range(len(file_data)):
    line = file_data[i]
    total += solve(line)
    print(i, total)
print("TOTAL:", total)
print("\nDIFF:", 170147128753455 - total)
assert total == 170147128753455

File: 2025/Day3/swift/D3PartTwo.swift
import Foundation

class D3PartTwo {
    func solve(line:String) -> Int {
        let lineInt = line.map{ Int(String($0)) }
        let totalBanks:Int = 12
        var res:[Int] = []
        var currentIdx = lineInt.startIndex
        var minVal:Int?
        for i in 0..<lineInt.count {
            if lineInt[i] == nil {
                continue
            }
            if res.count < totalBanks {
                if let temp = lineInt[i] {
                    res.append(temp)
                }
            }
            minVal = findMin(input: res)
            if minVal == nil {
                continue
            }
            if minVal! > lineInt[currentIdx]! {
                continue
            }
            //TODO: Determine which number to remove from the res to be able to add the current position to res
            var moveDown:Bool = false
            for i in 0..<res.count {
                if res[i] == minVal! && !moveDown{
                    moveDown = true
                    res[i] = res[i+1]
                } else if moveDown && i == res.count - 2 {
                    break
                } else if moveDown {
                    res[i] = res[i+1]
                }
            }
        }
        return res.reduce(0, {x, y in
            (x * 10) + y
        })
    }

    func findMin(input:[Int] = []) -> Int? {
        if input.count <= 0 {
            return nil
        }
        var min = input[0]
        for i in 1..<input.count {
            if min > input[i] {
                min = input[i]
            }
        }
        return min
    }

    func readFile(filename:String = "sample.txt") throws -> [String] {
        return try String(contentsOfFile: "../\(filename)", encoding: .utf8)
            .split(separator: "\n")
            .map { String($0) }
    }
}

let filename = "input.txt"
let d3p2 = D3PartTwo()
let fileData = try d3p2.readFile()
var resp:[Int] = []
for e in fileData {
    resp.append(d3p2.solve(line: e))
}
let result:Int = resp.reduce(0, {x, y in
    x + y
})
print("-> RESP: \(result)")